### PR TITLE
lxd/storage/drivers/btrfs: Fix optimized refresh

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by tomponline** [comment]
@monstermunchkin please can you update the tests to check for the failure scenario that was occurring before this fix, so that we can detect regressions in the future, as it seems there is a gap in the test suite that allowed this to be released.

**Comment by tomponline** [comment]
@monstermunchkin it would be good to add the tests first to recreate the error and then add the fix commits to this PR so we can see it resolve it.

File: test/suites/migration.sh [hunk 1]
```diff
@@ -424,6 +424,16 @@ migration() {
   # Test stateless copies
   lxc_remote copy --stateless l2:migratee/snap0 l1:migratee-new-name
 
+  if [ "${lxd_backend}" = "btrfs" ] && [ "${lxd2_backend}" = "btrfs" ] || [ "${lxd_backend}" = "zfs" ] && [ "${lxd2_backend}" = "zfs" ]; then
```
Comment (tomponline) [inline test/suites/migration.sh:427]: Why dont we do this test for all storage types (I realise only the btrfs and zfs ones will do optimized transfers, but would also test cross-remote copies for all storage drivers right) ?

File: test/suites/migration.sh [hunk 2]
```diff
@@ -424,6 +424,16 @@ migration() {
...
+    lxc_remote init testimage l1:c1
+    lxc_remote copy l1:c1 l2:c1
```
Comment (tomponline) [inline test/suites/migration.sh:429]: Suggest also starting the instance after copied to check it starts, and adding a file inside it to check that got copied to prove the copy process is working.

**Comment by stgraber** [comment]
@monstermunchkin the btrfs tests are failing

**Comment by monstermunchkin** [comment]
> @monstermunchkin the btrfs tests are failing

I'm working on it.

**Comment by tomponline** [comment]
@monstermunchkin given that the test suite has passed but there is no ZFS fix in this PR yet, it seems like the test suite is not yet testing for the ZFS failure scenario just yet.

**Review by tomponline — CHANGES_REQUESTED** [review]
Lets see the tests failing for ZFS before the fix goes in so we can be sure its thoroughly testing the failure scenario.

File: lxd/storage/drivers/driver_btrfs_volumes.go [hunk 3]
```diff
@@ -242,6 +242,13 @@ func (d *btrfs) CreateVolumeFromBackup(vol Volume, srcBackup backup.Info, srcDat
 		return "", fmt.Errorf("Could not find %q", srcFile)
 	}
 
+	type copyOp struct {
```
[the remark this answers is not in the document]
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_btrfs_volumes.go:245]: This is btrfs specific.

**Pull request merged**